File: ios/Sources/MessageBusPlugin/MessageBusPlugin.swift
import Foundation
import Capacitor

extension Notification.Name {
    static let capMessageBusPublish = Notification.Name("capMessageBusPublish")
}

@objc(MessageBusPlugin)
public class MessageBusPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "MessageBusPlugin"
    public let jsName = "MessageBus"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "sendMessage", returnType: CAPPluginReturnPromise)
    ]

    private var globalSub: MessageBus.Subscription?

    public override func load() {
        super.load()

        // Forward every native publish to JS
        globalSub = MessageBus.shared.subscribeAll { [weak self] type, payload in
            DispatchQueue.main.async {
                self?.notifyListeners("message", data: [
                    "type": type,
                    "payload": payload ?? NSNull()
                ])
            }
        }

        // NEW: listen for app-level posts and inject into the bus
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleExternalPublish(_:)),
            name: .capMessageBusPublish,
            object: nil
        )
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }

    @objc private func handleExternalPublish(_ note: Notification) {
        let userInfo = note.userInfo as? [String: Any]
        let type = userInfo?["type"] as? String ?? "unknown"
        let payload = userInfo?["payload"]
        MessageBus.shared.publish(type: type, payload: payload)
    }

    @objc func sendMessage(_ call: CAPPluginCall) {
        let type = call.getString("type") ?? "unknown"
        let payload = call.getObject("payload")

        // Publish to internal MessageBus (for JS listeners via globalSub)
        MessageBus.shared.publish(type: type, payload: payload)

        // Also post to NotificationCenter so native code can listen
        // This enables native modules like MediaPickers to receive JS messages
        NotificationCenter.default.post(
            name: .capMessageBusPublish,
            object: nil,
            userInfo: [
                "type": type,
                "payload": payload as Any
            ]
        )

        call.resolve(["ok": true])
    }
}

File: android/src/main/java/com/cnctd/plugins/messagebus/MessageBusPlugin.kt
package com.cnctd.plugins.messagebus

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

@CapacitorPlugin(name = "MessageBus")
class MessageBusPlugin : Plugin() {

    private var globalSub: MessageBus.Subscription? = null

    override fun load() {
        super.load()

        globalSub = MessageBus.shared.subscribeAll { type, payload ->
            activity.runOnUiThread {
                val data = JSObject().apply {
                    put("type", type)
                    when (payload) {
                        is JSObject -> put("payload", payload)
                        null -> put("payload", JSObject.NULL)
                        else -> put("payload", payload)
                    }
                }
                notifyListeners("message", data)
            }
        }
    }

    override fun handleOnDestroy() {
        super.handleOnDestroy()
        globalSub?.cancel()
        globalSub = null
    }

    @PluginMethod
    fun sendMessage(call: PluginCall) {
        val type = call.getString("type") ?: "unknown"
        val payload = call.getObject("payload")

        MessageBus.shared.publish(type, payload)

        call.resolve(JSObject().apply { put("ok", true) })
    }
}
